File: packages/go/qorechain/messages/cosmos.go
package messages

import (
	"time"

	"cosmossdk.io/math"
	feegranttypes "cosmossdk.io/x/feegrant"
	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
	sdk "github.com/cosmos/cosmos-sdk/types"
	authztypes "github.com/cosmos/cosmos-sdk/x/authz"
	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
	govv1types "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
	ibctransfertypes "github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
	ibcclienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
)

// The composers below build the standard Cosmos SDK messages QoreChain inherits.
// They reuse the upstream sdk types directly so the type URLs are exactly the
// canonical /cosmos.* and /ibc.* values the chain expects.

// ---- bank ----

// BankComposers builds bank module messages.
type BankComposers struct{}

// Bank is the bank message composer group.
var Bank BankComposers

// Send builds /cosmos.bank.v1beta1.MsgSend.
func (BankComposers) Send(fromAddress, toAddress string, amount sdk.Coins) *banktypes.MsgSend {
	return &banktypes.MsgSend{FromAddress: fromAddress, ToAddress: toAddress, Amount: amount}
}

// MultiSend builds /cosmos.bank.v1beta1.MsgMultiSend.
func (BankComposers) MultiSend(inputs []banktypes.Input, outputs []banktypes.Output) *banktypes.MsgMultiSend {
	return &banktypes.MsgMultiSend{Inputs: inputs, Outputs: outputs}
}

// ---- staking ----

// StakingComposers builds staking module messages.
type StakingComposers struct{}

// Staking is the staking message composer group.
var Staking StakingComposers

// Delegate builds /cosmos.staking.v1beta1.MsgDelegate.
func (StakingComposers) Delegate(delegator, validator string, amount sdk.Coin) *stakingtypes.MsgDelegate {
	return &stakingtypes.MsgDelegate{DelegatorAddress: delegator, ValidatorAddress: validator, Amount: amount}
}

// Undelegate builds /cosmos.staking.v1beta1.MsgUndelegate.
func (StakingComposers) Undelegate(delegator, validator string, amount sdk.Coin) *stakingtypes.MsgUndelegate {
	return &stakingtypes.MsgUndelegate{DelegatorAddress: delegator, ValidatorAddress: validator, Amount: amount}
}

// BeginRedelegate builds /cosmos.staking.v1beta1.MsgBeginRedelegate.
func (StakingComposers) BeginRedelegate(delegator, srcValidator, dstValidator string, amount sdk.Coin) *stakingtypes.MsgBeginRedelegate {
	return &stakingtypes.MsgBeginRedelegate{DelegatorAddress: delegator, ValidatorSrcAddress: srcValidator, ValidatorDstAddress: dstValidator, Amount: amount}
}

// ---- distribution ----

// DistributionComposers builds distribution module messages.
type DistributionComposers struct{}

// Distribution is the distribution message composer group.
var Distribution DistributionComposers

// WithdrawDelegatorReward builds /cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward.
func (DistributionComposers) WithdrawDelegatorReward(delegator, validator string) *distrtypes.MsgWithdrawDelegatorReward {
	return &distrtypes.MsgWithdrawDelegatorReward{DelegatorAddress: delegator, ValidatorAddress: validator}
}

// SetWithdrawAddress builds /cosmos.distribution.v1beta1.MsgSetWithdrawAddress.
func (DistributionComposers) SetWithdrawAddress(delegator, withdraw string) *distrtypes.MsgSetWithdrawAddress {
	return &distrtypes.MsgSetWithdrawAddress{DelegatorAddress: delegator, WithdrawAddress: withdraw}
}

// FundCommunityPool builds /cosmos.distribution.v1beta1.MsgFundCommunityPool.
func (DistributionComposers) FundCommunityPool(depositor string, amount sdk.Coins) *distrtypes.MsgFundCommunityPool {
	return &distrtypes.MsgFundCommunityPool{Depositor: depositor, Amount: amount}
}

// ---- gov ----

// GovComposers builds gov (v1) module messages.
type GovComposers struct{}

// Gov is the gov message composer group.
var Gov GovComposers

// SubmitProposal builds /cosmos.gov.v1.MsgSubmitProposal.
func (GovComposers) SubmitProposal(messages []*codectypes.Any, initialDeposit sdk.Coins, proposer, metadata, title, summary string) *govv1types.MsgSubmitProposal {
	return &govv1types.MsgSubmitProposal{Messages: messages, InitialDeposit: initialDeposit, Proposer: proposer, Metadata: metadata, Title: title, Summary: summary}
}

// Vote builds /cosmos.gov.v1.MsgVote.
func (GovComposers) Vote(proposalID uint64, voter string, option govv1types.VoteOption, metadata string) *govv1types.MsgVote {
	return &govv1types.MsgVote{ProposalId: proposalID, Voter: voter, Option: option, Metadata: metadata}
}

// Deposit builds /cosmos.gov.v1.MsgDeposit.
func (GovComposers) Deposit(proposalID uint64, depositor string, amount sdk.Coins) *govv1types.MsgDeposit {
	return &govv1types.MsgDeposit{ProposalId: proposalID, Depositor: depositor, Amount: amount}
}

// ---- authz ----

// AuthzComposers builds authz module messages.
type AuthzComposers struct{}

// Authz is the authz message composer group.
var Authz AuthzComposers

// Grant builds /cosmos.authz.v1beta1.MsgGrant.
func (AuthzComposers) Grant(granter, grantee string, grant authztypes.Grant) *authztypes.MsgGrant {
	return &authztypes.MsgGrant{Granter: granter, Grantee: grantee, Grant: grant}
}

// Revoke builds /cosmos.authz.v1beta1.MsgRevoke.
func (AuthzComposers) Revoke(granter, grantee, msgTypeURL string) *authztypes.MsgRevoke {
	return &authztypes.MsgRevoke{Granter: granter, Grantee: grantee, MsgTypeUrl: msgTypeURL}
}

// Exec builds /cosmos.authz.v1beta1.MsgExec.
func (AuthzComposers) Exec(grantee string, msgs []*codectypes.Any) *authztypes.MsgExec {
	return &authztypes.MsgExec{Grantee: grantee, Msgs: msgs}
}

// ---- feegrant ----

// FeegrantComposers builds feegrant module messages.
type FeegrantComposers struct{}

// Feegrant is the feegrant message composer group.
var Feegrant FeegrantComposers

// GrantAllowance builds /cosmos.feegrant.v1beta1.MsgGrantAllowance.
func (FeegrantComposers) GrantAllowance(granter, grantee string, allowance *codectypes.Any) *feegranttypes.MsgGrantAllowance {
	return &feegranttypes.MsgGrantAllowance{Granter: granter, Grantee: grantee, Allowance: allowance}
}

// RevokeAllowance builds /cosmos.feegrant.v1beta1.MsgRevokeAllowance.
func (FeegrantComposers) RevokeAllowance(granter, grantee string) *feegranttypes.MsgRevokeAllowance {
	return &feegranttypes.MsgRevokeAllowance{Granter: granter, Grantee: grantee}
}

// ---- ibc ----

// IBCComposers builds IBC application messages.
type IBCComposers struct{}

// IBC is the IBC message composer group.
var IBC IBCComposers

// Transfer builds /ibc.applications.transfer.v1.MsgTransfer (ICS-20).
//
// timeoutHeight may be the zero value for no height timeout; timeoutTimestamp is
// nanoseconds since the Unix epoch (0 for none). memo is optional.
func (IBCComposers) Transfer(sourcePort, sourceChannel string, token sdk.Coin, sender, receiver string, timeoutHeight ibcclienttypes.Height, timeoutTimestamp uint64, memo string) *ibctransfertypes.MsgTransfer {
	return &ibctransfertypes.MsgTransfer{
		SourcePort:       sourcePort,
		SourceChannel:    sourceChannel,
		Token:            token,
		Sender:           sender,
		Receiver:         receiver,
		TimeoutHeight:    timeoutHeight,
		TimeoutTimestamp: timeoutTimestamp,
		Memo:             memo,
	}
}

// BasicAllowance builds a feegrant BasicAllowance packed into an Any, ready for
// FeegrantComposers.GrantAllowance.
func BasicAllowance(spendLimit sdk.Coins, expiration *time.Time) (*codectypes.Any, error) {
	allowance := &feegranttypes.BasicAllowance{SpendLimit: spendLimit, Expiration: expiration}
	return codectypes.NewAnyWithValue(allowance)
}

var _ = math.ZeroInt

File: packages/rust/src/msg/cosmos.rs
//! Standard Cosmos SDK message composers (bank, staking, distribution, gov,
//! authz, feegrant, ibc) built from the `cosmrs`/`cosmos-sdk-proto` types, with
//! the canonical type URLs.
//!
//! These cover the messages a dApp commonly sends alongside the QoreChain custom
//! modules. Each returns a prost message; the `*_any` variants pack it into a
//! `cosmrs::Any`.

use crate::msg::to_any;
use cosmrs::proto::cosmos::base::v1beta1::Coin;
use cosmrs::Any;

use cosmrs::proto::cosmos::bank::v1beta1::MsgSend;
use cosmrs::proto::cosmos::distribution::v1beta1::{
    MsgWithdrawDelegatorReward, MsgWithdrawValidatorCommission,
};
use cosmrs::proto::cosmos::staking::v1beta1::{MsgBeginRedelegate, MsgDelegate, MsgUndelegate};

// --- bank ---

/// `/cosmos.bank.v1beta1.MsgSend` type URL.
pub const MSG_SEND: &str = "/cosmos.bank.v1beta1.MsgSend";

/// Builds the standard bank `MsgSend`.
pub fn bank_send(
    from_address: impl Into<String>,
    to_address: impl Into<String>,
    amount: Vec<Coin>,
) -> MsgSend {
    MsgSend {
        from_address: from_address.into(),
        to_address: to_address.into(),
        amount,
    }
}

/// Builds `MsgSend` packed into an `Any`.
pub fn bank_send_any(
    from_address: impl Into<String>,
    to_address: impl Into<String>,
    amount: Vec<Coin>,
) -> Any {
    to_any(&bank_send(from_address, to_address, amount), MSG_SEND)
}

// --- staking ---

/// `/cosmos.staking.v1beta1.MsgDelegate` type URL.
pub const MSG_DELEGATE: &str = "/cosmos.staking.v1beta1.MsgDelegate";
/// `/cosmos.staking.v1beta1.MsgUndelegate` type URL.
pub const MSG_UNDELEGATE: &str = "/cosmos.staking.v1beta1.MsgUndelegate";
/// `/cosmos.staking.v1beta1.MsgBeginRedelegate` type URL.
pub const MSG_BEGIN_REDELEGATE: &str = "/cosmos.staking.v1beta1.MsgBeginRedelegate";

/// Builds the staking `MsgDelegate`.
pub fn delegate(
    delegator_address: impl Into<String>,
    validator_address: impl Into<String>,
    amount: Coin,
) -> MsgDelegate {
    MsgDelegate {
        delegator_address: delegator_address.into(),
        validator_address: validator_address.into(),
        amount: Some(amount),
    }
}

/// Builds `MsgDelegate` packed into an `Any`.
pub fn delegate_any(
    delegator_address: impl Into<String>,
    validator_address: impl Into<String>,
    amount: Coin,
) -> Any {
    to_any(
        &delegate(delegator_address, validator_address, amount),
        MSG_DELEGATE,
    )
}

/// Builds the staking `MsgUndelegate`.
pub fn undelegate(
    delegator_address: impl Into<String>,
    validator_address: impl Into<String>,
    amount: Coin,
) -> MsgUndelegate {
    MsgUndelegate {
        delegator_address: delegator_address.into(),
        validator_address: validator_address.into(),
        amount: Some(amount),
    }
}

/// Builds `MsgUndelegate` packed into an `Any`.
pub fn undelegate_any(
    delegator_address: impl Into<String>,
    validator_address: impl Into<String>,
    amount: Coin,
) -> Any {
    to_any(
        &undelegate(delegator_address, validator_address, amount),
        MSG_UNDELEGATE,
    )
}

/// Builds the staking `MsgBeginRedelegate`.
pub fn begin_redelegate(
    delegator_address: impl Into<String>,
    validator_src_address: impl Into<String>,
    validator_dst_address: impl Into<String>,
    amount: Coin,
) -> MsgBeginRedelegate {
    MsgBeginRedelegate {
        delegator_address: delegator_address.into(),
        validator_src_address: validator_src_address.into(),
        validator_dst_address: validator_dst_address.into(),
        amount: Some(amount),
    }
}

/// Builds `MsgBeginRedelegate` packed into an `Any`.
pub fn begin_redelegate_any(
    delegator_address: impl Into<String>,
    validator_src_address: impl Into<String>,
    validator_dst_address: impl Into<String>,
    amount: Coin,
) -> Any {
    to_any(
        &begin_redelegate(
            delegator_address,
            validator_src_address,
            validator_dst_address,
            amount,
        ),
        MSG_BEGIN_REDELEGATE,
    )
}

// --- distribution ---

/// `/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward` type URL.
pub const MSG_WITHDRAW_DELEGATOR_REWARD: &str =
    "/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward";
/// `/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission` type URL.
pub const MSG_WITHDRAW_VALIDATOR_COMMISSION: &str =
    "/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission";

/// Builds the distribution `MsgWithdrawDelegatorReward`.
pub fn withdraw_delegator_reward(
    delegator_address: impl Into<String>,
    validator_address: impl Into<String>,
) -> MsgWithdrawDelegatorReward {
    MsgWithdrawDelegatorReward {
        delegator_address: delegator_address.into(),
        validator_address: validator_address.into(),
    }
}

/// Builds `MsgWithdrawDelegatorReward` packed into an `Any`.
pub fn withdraw_delegator_reward_any(
    delegator_address: impl Into<String>,
    validator_address: impl Into<String>,
) -> Any {
    to_any(
        &withdraw_delegator_reward(delegator_address, validator_address),
        MSG_WITHDRAW_DELEGATOR_REWARD,
    )
}

/// Builds the distribution `MsgWithdrawValidatorCommission`.
pub fn withdraw_validator_commission(
    validator_address: impl Into<String>,
) -> MsgWithdrawValidatorCommission {
    MsgWithdrawValidatorCommission {
        validator_address: validator_address.into(),
    }
}

/// Builds `MsgWithdrawValidatorCommission` packed into an `Any`.
pub fn withdraw_validator_commission_any(validator_address: impl Into<String>) -> Any {
    to_any(
        &withdraw_validator_commission(validator_address),
        MSG_WITHDRAW_VALIDATOR_COMMISSION,
    )
}

// --- gov (v1) ---

/// `/cosmos.gov.v1.MsgVote` type URL.
pub const MSG_VOTE: &str = "/cosmos.gov.v1.MsgVote";

/// Builds the gov `MsgVote`. `option` is the numeric `VoteOption`
/// (1 = yes, 2 = abstain, 3 = no, 4 = no-with-veto).
pub fn vote(
    proposal_id: u64,
    voter: impl Into<String>,
    option: i32,
    metadata: impl Into<String>,
) -> cosmrs::proto::cosmos::gov::v1::MsgVote {
    cosmrs::proto::cosmos::gov::v1::MsgVote {
        proposal_id,
        voter: voter.into(),
        option,
        metadata: metadata.into(),
    }
}

/// Builds `MsgVote` packed into an `Any`.
pub fn vote_any(
    proposal_id: u64,
    voter: impl Into<String>,
    option: i32,
    metadata: impl Into<String>,
) -> Any {
    to_any(&vote(proposal_id, voter, option, metadata), MSG_VOTE)
}

// --- authz ---

/// `/cosmos.authz.v1beta1.MsgExec` type URL.
pub const MSG_EXEC: &str = "/cosmos.authz.v1beta1.MsgExec";
/// `/cosmos.authz.v1beta1.MsgRevoke` type URL.
pub const MSG_REVOKE: &str = "/cosmos.authz.v1beta1.MsgRevoke";

/// Builds the authz `MsgExec`, wrapping the inner messages for `grantee` to
/// execute on behalf of the granter.
pub fn authz_exec(
    grantee: impl Into<String>,
    msgs: Vec<Any>,
) -> cosmrs::proto::cosmos::authz::v1beta1::MsgExec {
    cosmrs::proto::cosmos::authz::v1beta1::MsgExec {
        grantee: grantee.into(),
        msgs,
    }
}

/// Builds `MsgExec` packed into an `Any`.
pub fn authz_exec_any(grantee: impl Into<String>, msgs: Vec<Any>) -> Any {
    to_any(&authz_exec(grantee, msgs), MSG_EXEC)
}

/// Builds the authz `MsgRevoke` for a granted `msg_type_url`.
pub fn authz_revoke(
    granter: impl Into<String>,
    grantee: impl Into<String>,
    msg_type_url: impl Into<String>,
) -> cosmrs::proto::cosmos::authz::v1beta1::MsgRevoke {
    cosmrs::proto::cosmos::authz::v1beta1::MsgRevoke {
        granter: granter.into(),
        grantee: grantee.into(),
        msg_type_url: msg_type_url.into(),
    }
}

/// Builds `MsgRevoke` packed into an `Any`.
pub fn authz_revoke_any(
    granter: impl Into<String>,
    grantee: impl Into<String>,
    msg_type_url: impl Into<String>,
) -> Any {
    to_any(&authz_revoke(granter, grantee, msg_type_url), MSG_REVOKE)
}

// --- feegrant ---

/// `/cosmos.feegrant.v1beta1.MsgRevokeAllowance` type URL.
pub const MSG_REVOKE_ALLOWANCE: &str = "/cosmos.feegrant.v1beta1.MsgRevokeAllowance";

/// Builds the feegrant `MsgRevokeAllowance`.
pub fn revoke_allowance(
    granter: impl Into<String>,
    grantee: impl Into<String>,
) -> cosmrs::proto::cosmos::feegrant::v1beta1::MsgRevokeAllowance {
    cosmrs::proto::cosmos::feegrant::v1beta1::MsgRevokeAllowance {
        granter: granter.into(),
        grantee: grantee.into(),
    }
}

/// Builds `MsgRevokeAllowance` packed into an `Any`.
pub fn revoke_allowance_any(granter: impl Into<String>, grantee: impl Into<String>) -> Any {
    to_any(&revoke_allowance(granter, grantee), MSG_REVOKE_ALLOWANCE)
}

// --- ibc transfer ---
//
// The ICS-20 `MsgTransfer` is not bundled here: `cosmos-sdk-proto` (the proto
// crate `cosmrs` re-exports) does not ship the IBC types, and `to_any` accepts
// any prost message, so a caller wanting IBC can build the message from the
// `ibc-proto` crate and pack it directly:
//
// ```ignore
// use qorechain::msg::to_any;
// let any = to_any(&ibc_proto::ibc::applications::transfer::v1::MsgTransfer { .. },
//                  "/ibc.applications.transfer.v1.MsgTransfer");
// ```
/// `/ibc.applications.transfer.v1.MsgTransfer` type URL, exposed so callers can
/// pack an `ibc-proto` `MsgTransfer` with [`crate::msg::to_any`].
pub const MSG_TRANSFER: &str = "/ibc.applications.transfer.v1.MsgTransfer";
